riscv: telink: add dual mode for b92 and buteo and CI.

haiwentelink opened the pull request:
dual mode switch commit , it will works on b92 and tl3218x

liucoldstar commented on config/telink/chip-module/Kconfig.defaults:
```diff
@@ -260,6 +260,21 @@ config NVS_LOOKUP_CACHE_SIZE
+config I2C_LED
+    bool "Control the i2 led slave use raw address"
+    default n if I2C
+    default n
```
第二个default n看着多余？

haiwentelink replied:
第二行应该是之前写的有问题，之前的想法是想用depend on I2C，我再梳理下。

liucoldstar commented on config/telink/chip-module/Kconfig.defaults:
```diff
@@ -260,6 +260,21 @@ config NVS_LOOKUP_CACHE_SIZE
+    default n
+
+config DUAL_MODE_SWTICH
+    bool "Control the dual mode switch part"
```
建议改成bool "Enable Dual-Mode Switching"

haiwentelink replied:
config的变量不能加Enable，否则CI会报错。

liucoldstar replied:
我意思是把这个配置项的描述改成bool "Enable Dual-Mode Switching"

liucoldstar commented on examples/lighting-app/telink/src/AppTask.cpp:
```diff
@@ -60,6 +60,17 @@ CHIP_ERROR AppTask::Init(void)
+#if CONFIG_DUAL_MODE_SWTICH
+    if (sBoot_zb)
+    {
+        /* Switch from the touch link, need to restore previous values */
```
建议改成 /* Switching from TouchLink (Zigbee) to Matter. Restore previous states. */

liucoldstar commented on examples/lighting-app/telink/src/AppTask.cpp:
```diff
@@ -60,6 +60,17 @@ CHIP_ERROR AppTask::Init(void)
+        sfixture_on = user_para.onoff;
+        sBrightness = user_para.lightness;
+        sAppTask.UpdateClusterState();
+        printk("Matter: Updated ZB On/Off state and brightness.\n");
```
可以改进为：
printk("Matter: Restored Zigbee On/Off and brightness states.\n");

liucoldstar commented on examples/platform/telink/common/src/AppTaskCommon.cpp:
```diff
@@ -206,8 +231,45 @@ void AppTaskCommon::PowerOnFactoryReset(void)
+    {
+        return;
+    }
+    /*If initialization of Dnss takes longer than 60 seconds, the device will reboot and revert to Zigbee mode*/
```
/*
     * If Dnss initialization takes longer than 60 seconds,
     * the device will reboot and revert to Zigbee mode.
     */
    printk("Matter: DnssTimer expired. Rebooting...\n");

liucoldstar commented on examples/platform/telink/common/src/AppTaskCommon.cpp:
```diff
@@ -206,8 +231,45 @@ void AppTaskCommon::PowerOnFactoryReset(void)
 {
+
+#if CONFIG_DUAL_MODE_SWTICH
+    /* Proc ota boot flag , and erase flag */
```
/* Read OTA boot flag and erase flag from user parameters */

liucoldstar commented on examples/platform/telink/common/src/AppTaskCommon.cpp:
```diff
@@ -206,8 +231,45 @@ void AppTaskCommon::PowerOnFactoryReset(void)
+    if (user_para.val == USER_ZB_SW_VAL)
+    {
+        sBoot_zb = 1;
+        /* Ensure lightness is at least 2 to avoid display error on HomePod Mini */
```
/* Ensure brightness is at least 2 to avoid display issues on HomePod Mini */

liucoldstar commented on examples/platform/telink/common/src/AppTaskCommon.cpp:
```diff
@@ -206,8 +231,45 @@ void AppTaskCommon::PowerOnFactoryReset(void)
+        {
+            user_para.lightness = 2;
+        }
+        /* Pass the value to the init part to avoid gaps in pwm_pool init */
```
/*
         * Pass the brightness value to the initialization code
         * to avoid gaps in PWM pool initialization.
         */

liucoldstar commented on examples/platform/telink/common/src/AppTaskCommon.cpp:
```diff
@@ -206,8 +231,45 @@ void AppTaskCommon::PowerOnFactoryReset(void)
+        }
+        k_timer_init(&sDnssTimer, &AppTask::DnssTimerTimeoutCallback, nullptr);
+        k_timer_start(&sDnssTimer, K_MSEC(kDnssTimeout), K_NO_WAIT);
+        printk("Matter: start timer to protect Dnss initialized %x \r\n", *(int *) (&user_para));
```
printk("Matter: Started DNS protection timer. user_para=%x\n", *(int *)(&user_para));

liucoldstar commented on examples/platform/telink/common/src/AppTaskCommon.cpp:
```diff
@@ -206,8 +231,45 @@ void AppTaskCommon::PowerOnFactoryReset(void)
+#if CONFIG_DUAL_MODE_SWTICH
+    /* Proc ota boot flag , and erase flag */
+    flash_read(flash_para_dev, USER_PARTITION_OFFSET, &user_para, sizeof(user_para));
+    /* Boot from Zigbee , need to clean the user parameters sector first and set a flag */
```
/*
     * If the device boots from Zigbee, set a flag and adjust user parameters.
     * Then start a timer to ensure Dnss initialization completes within
     * a given timeframe.
     */

liucoldstar commented on examples/platform/telink/common/src/AppTaskCommon.cpp:
```diff
@@ -588,6 +650,10 @@ void AppTaskCommon::FactoryResetHandler(AppEvent * aEvent)
         sFactoryResetCntr = 0;
 
+#if CONFIG_DUAL_MODE_SWTICH
+        printk("Factory reset triggered by button, resetting to Zigbee mode");
```
printk("Factory reset triggered by button; reverting to Zigbee mode\n");

liucoldstar commented on examples/platform/telink/common/src/AppTaskCommon.cpp:
```diff
@@ -745,6 +811,28 @@ void AppTaskCommon::ChipEventHandler(const ChipDeviceEvent * event, intptr_t /*
+        uint8_t val = USER_MATTER_PAIR_VAL;
+        flash_erase(flash_para_dev, USER_PARTITION_OFFSET, USER_PARTITION_SIZE);
+        flash_write(flash_para_dev, USER_PARTITION_OFFSET, &val, 1);
+        printk("Commissioning complete, set Matter commissionined flag");
```
printk("Commissioning complete; Matter commissioned flag set.\n");

liucoldstar commented on examples/platform/telink/common/src/AppTaskCommon.cpp:
```diff
@@ -745,6 +811,28 @@ void AppTaskCommon::ChipEventHandler(const ChipDeviceEvent * event, intptr_t /*
+        /* Reset to Zigbee mode if commissioning fails */
+        if (sBoot_zb)
+        {
+            printk("FailSafeTimer expired, Matter commissioning failed, rebooting to Zigbee mode.\r\n");
```
printk("FailSafeTimer expired; Matter commissioning failed. Rebooting to Zigbee mode...\n");

liucoldstar commented on examples/platform/telink/common/src/AppTaskCommon.cpp:
```diff
@@ -745,6 +811,28 @@ void AppTaskCommon::ChipEventHandler(const ChipDeviceEvent * event, intptr_t /*
+        }
+        else
+        {
+            printk("FailSafeTimer expired, Matter commissioning failed.\r\n");
```
printk("FailSafeTimer expired; Matter commissioning failed.\n");

damien0x0023 commented on config/telink/chip-module/Kconfig.defaults:
```diff
@@ -260,6 +260,21 @@ config NVS_LOOKUP_CACHE_SIZE
+
+config I2C_LED
+    bool "Control the i2 led slave use raw address"
+    default n if I2C
```
Why do we need two default NOs here? Can we set it to "default y if I2C".

damien0x0023 commented on config/telink/chip-module/Kconfig.defaults:
```diff
@@ -34,7 +34,7 @@ choice LOG_MODE
 
 choice MATTER_LOG_LEVEL_CHOICE
-    default MATTER_LOG_LEVEL_WRN if SOC_RISCV_TELINK_TL321X
+    default MATTER_LOG_LEVEL_DBG if SOC_RISCV_TELINK_TL321X
```
Could we remove this line as they all use DBG level?

haiwentelink replied:
@damien0x0023  For tl321x which will not have enough flash size ,open just for debug , or will close.but for 4m , which we will have enough space , reserver option for different soc .

damien0x0023 commented on examples/lighting-app/telink/src/AppTask.cpp:
```diff
@@ -60,6 +60,17 @@ CHIP_ERROR AppTask::Init(void)
+        sfixture_on = user_para.onoff;
+        sBrightness = user_para.lightness;
+        sAppTask.UpdateClusterState();
+        printk("Matter: Updated ZB On/Off state and brightness.\n");
```
Could we use Matter's logging API (ChipLogProgress) or Zephyr's logging API (LOG_INF or LOG_DBG) to replace printk()?

haiwentelink replied:
should not bound to the log info , we want this print take effect in all case , if we open the matter print , the log will lost .

liucoldstar commented on examples/platform/telink/common/src/AppTaskCommon.cpp:
```diff
@@ -758,6 +846,14 @@ void AppTaskCommon::ChipEventHandler(const ChipDeviceEvent * event, intptr_t /*
+        if (sBoot_zb)
+        {
+            k_timer_stop(&sDnssTimer);
+            printk("Dnss Timer stopped, Matter commissioning kDnssdInitialized.\r\n");
```
printk("DnssTimer stopped; DNS-SD has been initialized.\n");

damien0x0023 commented on examples/platform/telink/common/src/AppTaskCommon.cpp:
```diff
@@ -159,6 +179,11 @@ class AppFabricTableDelegate : public FabricTable::Delegate
                     ChipLogError(DeviceLayer, "Storage clearance failed: %d", status);
                 }
+#if CONFIG_DUAL_MODE_SWTICH
+                printk("Erasing user parameters and resetting to Zigbee mode");
```
I have the same concern about logging API here.

haiwentelink replied:
should not bound to the log info , we want this print take effect in all case , if we open the matter print , the log will lost .

liucoldstar commented on examples/platform/telink/common/src/AppTaskCommon.cpp:
```diff
@@ -67,6 +73,12 @@ constexpr uint32_t kIdentifyFinishOffRateMs     = 50;
+#if CONFIG_DUAL_MODE_SWTICH
+const struct device * flash_para_dev = USER_PARTITION_DEVICE;
+constexpr int kDnssTimeout           = 60000;
+k_timer sDnssTimer;
```
建议加static，因为sDnsTimer只在这个文件用到了。

haiwentelink replied:
ok

liucoldstar commented on src/platform/telink/tl3218x_2m_flash.overlay:
```diff
@@ -9,31 +9,55 @@
 		boot_partition: partition@0 {
 			label = "mcuboot";
-			reg = <0x00000000 0x13000>;
+			reg = <0x00000000 0x12000>;// bootloader should use the size optimize to low down to 72k,reserve 4k for user.
```
boot_partition: partition@0 {
    label = "mcuboot";
    // The bootloader is optimized to fit within 72 KB (0x12000).
    // 4 KB (0x1000) are reserved for user data or future expansion.
    reg = <0x00000000 0x12000>;
};

liucoldstar commented on src/platform/telink/tl3218x_2m_flash.overlay:
```diff
@@ -9,31 +9,55 @@
+		slot0_partition: partition@12000 {
 			label = "image-0";
-			reg = <0x13000 0xef000>;
+			reg = <0x12000 0x122000>;// 0x12000 is matter, and reserve 912k for matter , 0xF6000 is zb reserve 248k for zigbee
```
slot0_partition: partition@12000 {
    label = "image-0";
    // Starting at 0x12000, this partition reserves 912 KB for Matter,
    // and an additional 248 KB for Zigbee, totaling 0x122000 in size.
    reg = <0x12000 0x122000>;
};

liucoldstar commented on src/platform/telink/tl3218x_2m_flash.overlay:
```diff
@@ -9,31 +9,55 @@
-		factory_rfu_partition: partition@102800 {
-			label = "factory-data-rfu";
-			reg = <0x102800 0x800>;
+		slot1_partition: partition@134000 {
```
slot1_partition: partition@134000 {
    label = "image-1";
    // Uses LZMA compression to reduce binary size.
    // The combined space (920 KB + 248 KB) is reduced to ~60% (~712 KB).
    // Zigbee NVS data will also reside here.
    reg = <0x134000 0xb2000>;
};

liucoldstar commented on src/platform/telink/tl3218x_2m_flash.overlay:
```diff
@@ -9,31 +9,55 @@
+			reg = <0x134000 0xb2000>;// the slot1 is use LZMA to save binsize . (920k + 248k)*60% = 712k, zb nvs will locate here
 		};
-		storage_partition: partition@103000 {
+		storage_partition: partition@1e6000 {
```
storage_partition: partition@1e6000 {
    label = "storage";
    // NVS storage for Matter.
    reg = <0x1e6000 0xb000>;
};

liucoldstar commented on src/platform/telink/tl3218x_2m_flash.overlay:
```diff
@@ -9,31 +9,55 @@
-		slot1_partition: partition@10f000 {
-			label = "image-1";
-			reg = <0x10f000 0xef000>;
+		user_token_partition: partition@1f1000 {
```
user_token_partition: partition@1f1000 {
    label = "user-token";
    // Stores token information. If unused, this partition can be removed.
    reg = <0x1f1000 0x1000>;
};

liucoldstar commented on src/platform/telink/tl3218x_2m_flash.overlay:
```diff
@@ -9,31 +9,55 @@
+ 			label = "user-token";
+			reg = <0x1f1000 0x1000>; //user:store token info, if don't use ,can delete
+ 		};
+		user_para_partition: partition@1f2000 {
```
user_para_partition: partition@1f2000 {
    label = "user-para";
    // Stores mode information and key data.
    reg = <0x1f2000 0x1000>;
};

damien0x0023 reviewed (commented):
@haiwentelink 
I pulled all your working branches in matter/zephyr/hal_telink/mcuboot and built dual-mode firmware on tlsr9528a/tl3218a. Both of them work with standalone ZigBee firmware. 
However, I have some concerns about wording and logging. Please check if we can refine them.

liucoldstar commented on src/platform/telink/tl3218x_2m_flash.overlay:
```diff
@@ -9,31 +9,55 @@
+ 		};
+		secure_partition: partition@1f8000 {
+ 			label = "secure";
+			reg = <0x1f8000 0x1000>; //secure info ,reserved for secure boot .if not use , can beused by other way .
```
secure_partition: partition@1f8000 {
    label = "secure";
    // Secure area reserved for secure boot.
    // If not in use, it may be repurposed.
    reg = <0x1f8000 0x1000>;
};

liucoldstar commented on src/platform/telink/tl3218x_2m_flash.overlay:
```diff
@@ -9,31 +9,55 @@
+ 		};
+		factory_partition: partition@1f9000 {
+			label = "factory-data";
+			reg = <0x1f9000 0x1000>; // factory data info and dac info
```
// Factory data

liucoldstar commented on src/platform/telink/tl3218x_2m_flash.overlay:
```diff
@@ -9,31 +9,55 @@
+		};
+		dac_keypair_partition: partition@1f9800 {
+			label = "dac-keypair";
+			reg = <0x1f9800 0x800>; //store dac and key pair.
```
dac_keypair_partition: partition@1f9800 {
    label = "dac-keypair";
    // Stores DAC certificate and key pair (2 KB).
    reg = <0x1f9800 0x800>;
};

liucoldstar commented on src/platform/telink/tl3218x_2m_flash.overlay:
```diff
@@ -9,31 +9,55 @@
 		};
+		user_rfu_partition: partition@1fa000 {
+ 			label = "user_rfu";
+			reg = <0x1fa000 0x4000>; //user_rfu: reserve 16k for user extend.
```
// Reserves 16 KB for user-specific extensions.

liucoldstar commented on src/platform/telink/tl3218x_2m_flash.overlay:
```diff
@@ -9,31 +9,55 @@
 		vendor_partition: partition@1fe000 {
 			label = "vendor-data";
-			reg = <0x1fe000 0x2000>;
+			reg = <0x1fe000 0x2000>;// mac and adc info.
```
// Holds MAC, ADC, or other vendor-specific data (8 KB).

haiwentelink committed:
riscv: telink: add dual mode for b92 and buteo and CI.

- clean the boot flag in init part .
- add the kDnsstimer to protect from init to dnss .
- add special proc for ikea.
- add the user-para proc in matter.
- add 2m flash for buteo ,and 4m for b92.
- update CI for b92 and tl3218x .
- fix the switch logic part .

Signed-off-by: Haiwen Xia <haiwen.xia@telink-semi.com>
